core/src/commonMain/kotlin/ua/syt0r/kanji/core/stroke_evaluator/AltKanjiStrokeEvaluator.kt

```diff
@@ -16,11 +18,71 @@ import kotlin.math.sqrt
 class AltKanjiStrokeEvaluator : KanjiStrokeEvaluator {
 
     companion object {
+        // difficulty. lower is easier.
+        private const val DIFFICULTY = 1.0f
+
         private const val SIMILARITY_ERROR_THRESHOLD = 10f
-        private const val MAX_INTERPOLATION_POINTS = 22
-        private const val MIN_INTERPOLATION_POINTS = 5
-        private const val SEGMENT_SCALE = 5
-        private const val DEAD_BAND = 22f
+
+        // divide path into segments of this length
+        private const val SEGMENT_LENGTH = 5f
+
+        // dead band below which there is no error contribution
+        private const val POSITIONAL_DEAD_BAND = 44f / DIFFICULTY
+
+        // dead band (in degrees) below which there is no error contribution
+        private const val DIRECTIONAL_DEAD_BAND = 5f / DIFFICULTY
+
+        // skipping path segments of either path to get a better match still incurs some error
+        // see function gapCost()
+        private const val GAP_OPENING = 5f
+        private const val GAP_EXTENSION = 5f
+
+        // max positional error in number of dead bands
+        private const val MAX_POSITIONAL_ERROR = 6f / DIFFICULTY
+
+        // max directional error in degrees
+        private const val MAX_DIRECTIONAL_ERROR = 30f / DIFFICULTY
+
+        // scaling constant to make balancing with gap costs easier
+        private const val ERROR_SCALE = 20f
+    }
+
+    private data class Vector2d(val a: Double, val b: Double) : Comparable<Vector2d> {
+        operator fun plus(other: Vector2d) = Vector2d(a + other.a, b + other.b)
+
+        // adding equal amounts of scalar to both components
+        operator fun plus(scalar: Double): Vector2d {
+            if ((a == 0.0) && (b == 0.0)) {
+                val ab = sqrt(0.5 * (scalar.pow(2.0)))
+                return Vector2d(ab, ab)
+            }
+
+            if (a == 0.0) return Vector2d(0.0, b + scalar)
+
+            // the following handles all other cases including b==0.0
+            val newLength = this.length() + scalar
+            val denominator = sqrt((b / a).pow(2.0) + 1)
+
+            return Vector2d(newLength / denominator, newLength * (b / a) / denominator)
+        }
+
+        operator fun div(scalar: Double):Vector2d = Vector2d(this.a/scalar,this.b/scalar)
+
+        // there's no order on complex numbers or 2D vectors. We do it a bit wonky here.
+        // correct? no! sufficient? yes!
+        override operator fun compareTo(other: Vector2d): Int {
```

Comment:
Any reason to have this? Are you sorting this class in list somewhere? Also, is this class really needed? I see you're using more precision here, but is there really any real impact? I think you could have used previous `PointF` and just create private extensions something like `private fun PointF.plus(scalar...): PointF` or similar

Reply:
Comparison is needed for backtracing through the matrix to get the actual path length. I didn't want to reuse your PointF class because it seems to have another intention. Sure, both classes just hold a pair of values but one class represents coordinates and the other some obscure two-dimensional error values. If (in the future) you decide to change your class to do something more or do something differently, it will create unnecessary dependencies. That's why I just did my own class.

Reply:
Alright, let's leave it